README.md
# oracledb-upsert

[![NPM Version](http://img.shields.io/npm/v/oracledb-upsert.svg?style=flat)](https://www.npmjs.org/package/oracledb-upsert) [![CI](https://github.com/sagiegurari/oracledb-upsert/workflows/CI/badge.svg?branch=master)](https://github.com/sagiegurari/oracledb-upsert/actions) [![Coverage Status](https://coveralls.io/repos/sagiegurari/oracledb-upsert/badge.svg)](https://coveralls.io/r/sagiegurari/oracledb-upsert) [![Known Vulnerabilities](https://snyk.io/test/github/sagiegurari/oracledb-upsert/badge.svg)](https://snyk.io/test/github/sagiegurari/oracledb-upsert) [![Inline docs](http://inch-ci.org/github/sagiegurari/oracledb-upsert.svg?branch=master)](http://inch-ci.org/github/sagiegurari/oracledb-upsert) [![License](https://img.shields.io/npm/l/oracledb-upsert.svg?style=flat)](https://github.com/sagiegurari/oracledb-upsert/blob/master/LICENSE) [![Total Downloads](https://img.shields.io/npm/dt/oracledb-upsert.svg?style=flat)](https://www.npmjs.org/package/oracledb-upsert)

> UPSERT (insert/update) extension to oracledb.

* [Overview](#overview)
* [Usage](#usage)
* [Installation](#installation)
* [API Documentation](docs/api.md)
* [Contributing](.github/CONTRIBUTING.md)
* [Release History](#history)
* [License](#license)

<a name="overview"></a>
## Overview
This library extends the oracledb connection object with a new upsert function to enable to run UPDATE or INSERT based on the
data currently in the DB.

<a name="usage"></a>
## Usage
In order to use this library, you need to extend the main oracledb object as follows:

```js
//load the oracledb library
var oracledb = require('oracledb');

//load the simple oracledb
var SimpleOracleDB = require('simple-oracledb');

//modify the original oracledb library
SimpleOracleDB.extend(oracledb);

//load the extension
require('oracledb-upsert');

//from this point connections fetched via oracledb.getConnection(...) or pool.getConnection(...)
//have access to the UPSERT function.
oracledb.getConnection(function onConnection(error, connection) {
    if (error) {
        //handle error
    } else {
        //work with new capabilities or original oracledb capabilities
        connection.upsert({
            query: 'SELECT ID FROM MY_DATA WHERE ID = :id',
            insert: 'INSERT INTO MY_DATA (ID, NAME) VALUES (:id, :name)',
            update: 'UPDATE MY_DATA SET NAME = :name WHERE ID = :id'
        }, {
            id: 110,
            name: 'new name'
        }, {
            autoCommit: false
        }, function onUpsert(error, results) {
            if (error) {
                //handle error...
            } else {
                console.log('rows affected: ', results.rowsAffected);

                //continue flow...
            }
        });
    }
});
```

<a name="connection-upsert"></a>
<!-- markdownlint-disable MD009 MD031 MD036 -->
### 'connection.upsert(sqls, bindParams, [options], [callback]) ⇒ [Promise]'
The UPSERT oracledb extension gets 3 SQL statements.<br>
It first queries the database for existing data, based on the output, it either runs INSERT or UPDATE SQL.<br>
If it runs the INSERT and it fails on unique constraint, it will also run the UPDATE.<br>
The output in the callback is the output of the INSERT/UPDATE operation.

**Example**  
```js
connection.upsert({
  query: 'SELECT ID FROM MY_DATA WHERE ID = :id',
  insert: 'INSERT INTO MY_DATA (ID, NAME) VALUES (:id, :name)',
  update: 'UPDATE MY_DATA SET NAME = :name WHERE ID = :id'
}, {
  id: 110,
  name: 'new name'
}, {
  autoCommit: false
}, function onUpsert(error, results) {
  if (error) {
    //handle error...
  } else {
    console.log('rows affected: ', results.rowsAffected);

    //continue flow...
  }
});
```
<!-- markdownlint-enable MD009 MD031 MD036 -->

<a name="installation"></a>
## Installation
In order to use this library, just run the following npm install command:

```sh
npm install --save oracledb-upsert
```

This library doesn't define oracledb as a dependency and therefore it is not installed when installing oracledb-upsert.<br>
You should define oracledb in your package.json and install it based on the oracledb installation instructions found at: [installation guide](https://github.com/oracle/node-oracledb/blob/master/INSTALL.md)

## API Documentation
See full docs at: [API Docs](docs/api.md)

## Contributing
See [contributing guide](.github/CONTRIBUTING.md)

<a name="history"></a>
## Release History

| Date        | Version | Description |
| ----------- | ------- | ----------- |
| 2020-05-13  | v2.0.0  | Migrate to github actions and upgrade minimal node version |
| 2019-02-08  | v1.2.1  | Maintenance |
| 2016-08-05  | v0.0.43 | Added promise support |
| 2016-03-04  | v0.0.1  | Initial release. |

<a name="license"></a>
## License
Developed by Sagie Gur-Ari and licensed under the Apache 2 open source license.


docs/CHANGELOG.md
| Date        | Version | Description |
| ----------- | ------- | ----------- |
| 2020-05-13  | v2.0.0  | Migrate to github actions and upgrade minimal node version |
| 2019-02-08  | v1.2.1  | Maintenance |
| 2016-08-05  | v0.0.43 | Added promise support |
| 2016-03-04  | v0.0.1  | Initial release. |


test/spec/upsert-spec.js
'use strict';

const chai = require('chai');
const assert = chai.assert;
const PromiseLib = global.Promise || require('promiscuous');
const SimpleOracleDB = require('simple-oracledb');
require('../../lib/upsert');

describe('upsert Tests', function () {
    const asArray = function (args) {
        return Array.prototype.slice.call(args, 0);
    };

    const createOracleDB = function () {
        const failFunc = function () {
            assert.fail();
        };

        const oracledb = {
            createPool() {
                return undefined;
            },
            getConnection() {
                const args = asArray(arguments);
                const callback = args[args.length - 1];
                callback(null, {
                    query: failFunc,
                    insert: failFunc,
                    update: failFunc
                });
            }
        };

        SimpleOracleDB.extend(oracledb);

        return oracledb;
    };

    describe('extension', function () {
        it('valid', function (done) {
            const oracledb = createOracleDB();

            oracledb.getConnection({}, function (error, connection) {
                assert.isNull(error);
                assert.isFunction(connection.upsert);

                done();
            });
        });
    });

    describe('upsert', function () {
        it('no options', function (done) {
            const oracledb = createOracleDB();

            oracledb.getConnection({}, function (connectionError, connection) {
                assert.isNull(connectionError);

                connection.query = function () {
                    const args = asArray(arguments);
                    args[args.length - 1](new Error('test error'));
                };

                connection.upsert({}, {}, function (error, result) {
                    assert.isDefined(error);
                    assert.equal(error.message, 'test error');
                    assert.isUndefined(result);

                    done();
                });
            });
        });

        it('null bind params', function (done) {
            const oracledb = createOracleDB();

            oracledb.getConnection({}, function (connectionError, connection) {
                assert.isNull(connectionError);

                connection.query = function () {
                    arguments[arguments.length - 1](new Error('test error'));
                };

                connection.upsert({}, null, function (error, result) {
                    assert.isDefined(error);
                    assert.equal(error.message, 'test error');
                    assert.isUndefined(result);

                    done();
                });
            });
        });

        it('array bind params', function (done) {
            const oracledb = createOracleDB();

            oracledb.getConnection({}, function (connectionError, connection) {
                assert.isNull(connectionError);

                connection.upsert({}, [], function (error, result) {
                    assert.isDefined(error);
                    assert.equal(error.message, 'Array type bind params are not supported.');
                    assert.isUndefined(result);

                    done();
                });
            });
        });

        it('query error', function (done) {
            const oracledb = createOracleDB();

            oracledb.getConnection({}, function (connectionError, connection) {
                assert.isNull(connectionError);

                connection.query = function () {
                    arguments[arguments.length - 1](new Error('test query error'));
                };

                connection.upsert({}, {}, {}, function (error, result) {
                    assert.isDefined(error);
                    assert.equal(error.message, 'test query error');
                    assert.isUndefined(result);

                    done();
                });
            });
        });

        it('insert general error', function (done) {
            const oracledb = createOracleDB();

            oracledb.getConnection({}, function (connectionError, connection) {
                assert.isNull(connectionError);

                connection.query = function () {
                    arguments[arguments.length - 1](null, []);
                };

                connection.insert = function () {
                    arguments[arguments.length - 1](new Error('test insert error'));
                };

                connection.upsert({}, {}, {}, function (error, result) {
                    assert.isDefined(error);
                    assert.equal(error.message, 'test insert error');
                    assert.isUndefined(result);

                    done();
                });
            });
        });

        it('insert unique constraint error and update error', function (done) {
            const oracledb = createOracleDB();

            oracledb.getConnection({}, function (connectionError, connection) {
                assert.isNull(connectionError);

                connection.query = function () {
                    arguments[arguments.length - 1](null, []);
                };

                connection.insert = function () {
                    arguments[arguments.length - 1](new Error('test ORA-00001 error'));
                };

                connection.update = function () {
                    arguments[arguments.length - 1](new Error('test update error'));
                };

                connection.upsert({}, {}, {}, function (error, result) {
                    assert.isDefined(error);
                    assert.equal(error.message, 'test update error');
                    assert.isUndefined(result);

                    done();
                });
            });
        });

        it('no insert and update error', function (done) {
            const oracledb = createOracleDB();

            oracledb.getConnection({}, function (connectionError, connection) {
                assert.isNull(connectionError);

                connection.query = function () {
                    arguments[arguments.length - 1](null, [{}]);
                };

                connection.update = function () {
                    arguments[arguments.length - 1](new Error('test update2 error'));
                };

                connection.upsert({}, {}, {}, function (error, result) {
                    assert.isDefined(error);
                    assert.equal(error.message, 'test update2 error');
                    assert.isUndefined(result);

                    done();
                });
            });
        });

        it('no insert and update error, using promise', function (done) {
            const oracledb = createOracleDB();

            oracledb.getConnection({}, function (connectionError, connection) {
                assert.isNull(connectionError);

                connection.query = function () {
                    arguments[arguments.length - 1](null, [{}]);
                };

                connection.update = function () {
                    arguments[arguments.length - 1](new Error('test update2 error'));
                };

                global.Promise = PromiseLib;

                const promise = connection.upsert({}, {}, {});

                promise.then(function () {
                    assert.fail();
                }).catch(function (error) {
                    assert.isDefined(error);
                    assert.equal(error.message, 'test update2 error');

                    done();
                });
            });
        });

        it('insert valid', function (done) {
            const oracledb = createOracleDB();

            oracledb.getConnection({}, function (connectionError, connection) {
                assert.isNull(connectionError);

                connection.query = function () {
                    arguments[arguments.length - 1](null, []);
                };

                connection.insert = function () {
                    arguments[arguments.length - 1](null, {
                        rowsAffected: 1
                    });
                };

                connection.upsert({}, {}, {}, function (error, result) {
                    assert.isNull(error);
                    assert.deepEqual(result, {
                        rowsAffected: 1
                    });

                    done();
                });
            });
        });

        it('insert valid, using promise', function (done) {
            const oracledb = createOracleDB();

            oracledb.getConnection({}, function (connectionError, connection) {
                assert.isNull(connectionError);

                connection.query = function () {
                    arguments[arguments.length - 1](null, []);
                };

                connection.insert = function () {
                    arguments[arguments.length - 1](null, {
                        rowsAffected: 1
                    });
                };

                global.Promise = PromiseLib;

                const promise = connection.upsert({}, {}, {});

                promise.then(function (result) {
                    assert.deepEqual(result, {
                        rowsAffected: 1
                    });

                    done();
                }).catch(function () {
                    assert.fail();
                });
            });
        });

        it('row exists, update valid', function (done) {
            const oracledb = createOracleDB();

            oracledb.getConnection({}, function (connectionError, connection) {
                assert.isNull(connectionError);

                connection.query = function () {
                    arguments[arguments.length - 1](null, [{}]);
                };

                connection.update = function () {
                    arguments[arguments.length - 1](null, {
                        rowsAffected: 1
                    });
                };

                connection.upsert({
                    query: 'SELECT * FROM MY_TABLE WHERE A = :a',
                    update: 'UPDATE MY_TABLE SET B = :b, MY_CLOB = EMPTY_CLOB() WHERE A = :a'
                }, {
                    a: 1,
                    b: 2,
                    myClob: 3
                }, {
                    lobMetaInfo: {
                        MY_CLOB: 'myClob'
                    }
                }, function (error, result) {
                    assert.isNull(error);
                    assert.deepEqual(result, {
                        rowsAffected: 1
                    });

                    done();
                });
            });
        });

        it('row exists, update valid, using promise', function (done) {
            const oracledb = createOracleDB();

            oracledb.getConnection({}, function (connectionError, connection) {
                assert.isNull(connectionError);

                connection.query = function () {
                    arguments[arguments.length - 1](null, [{}]);
                };

                connection.update = function () {
                    arguments[arguments.length - 1](null, {
                        rowsAffected: 1
                    });
                };

                global.Promise = PromiseLib;

                const promise = connection.upsert({
                    query: 'SELECT * FROM MY_TABLE WHERE A = :a',
                    update: 'UPDATE MY_TABLE SET B = :b, MY_CLOB = EMPTY_CLOB() WHERE A = :a'
                }, {
                    a: 1,
                    b: 2,
                    myClob: 3
                }, {
                    lobMetaInfo: {
                        MY_CLOB: 'myClob'
                    }
                });

                promise.then(function (result) {
                    assert.deepEqual(result, {
                        rowsAffected: 1
                    });

                    done();
                }).catch(function () {
                    assert.fail();
                });
            });
        });

        it('row exists, update did not modify any row', function (done) {
            const oracledb = createOracleDB();

            oracledb.getConnection({}, function (connectionError, connection) {
                assert.isNull(connectionError);

                connection.query = function () {
                    arguments[arguments.length - 1](null, [{}]);
                };

                connection.update = function () {
                    arguments[arguments.length - 1](null, {
                        rowsAffected: 0
                    });
                };

                connection.upsert({}, {}, {}, function (error, result) {
                    assert.isDefined(error);
                    assert.equal(error.message, 'No rows updated.');
                    assert.isUndefined(result);

                    done();
                });
            });
        });

        it('insert unique constraint error and update valid', function (done) {
            const oracledb = createOracleDB();

            oracledb.getConnection({}, function (connectionError, connection) {
                assert.isNull(connectionError);

                connection.query = function () {
                    arguments[arguments.length - 1](null, []);
                };

                let insertCalled = false;
                connection.insert = function () {
                    insertCalled = true;

                    arguments[arguments.length - 1](new Error('test ORA-00001 error'));
                };

                connection.update = function () {
                    arguments[arguments.length - 1](null, {
                        rowsAffected: 1
                    });
                };

                connection.upsert({}, {}, {}, function (error, result) {
                    assert.isNull(error);
                    assert.isTrue(insertCalled);
                    assert.deepEqual(result, {
                        rowsAffected: 1
                    });

                    done();
                });
            });
        });

        it('insert no impact and update valid', function (done) {
            const oracledb = createOracleDB();

            oracledb.getConnection({}, function (connectionError, connection) {
                assert.isNull(connectionError);

                connection.query = function () {
                    const args = asArray(arguments);
                    assert.equal(args[0], 'select 1');

                    args[args.length - 1](null, []);
                };

                let insertCalled = false;
                connection.insert = function () {
                    const args = asArray(arguments);
                    assert.equal(args[0], 'insert 2');

                    insertCalled = true;

                    args[args.length - 1](null, {
                        rowsAffected: 0
                    });
                };

                connection.update = function () {
                    const args = asArray(arguments);
                    assert.equal(args[0], 'update 3');

                    args[args.length - 1](null, {
                        rowsAffected: 1
                    });
                };

                connection.upsert({
                    query: 'select 1',
                    insert: 'insert 2',
                    update: 'update 3'
                }, {}, {}, function (error, result) {
                    assert.isNull(error);
                    assert.isTrue(insertCalled);
                    assert.deepEqual(result, {
                        rowsAffected: 1
                    });

                    done();
                });
            });
        });
    });
});


Gruntfile.js
'use strict';

/*jslint nomen: true*/
//jscs:disable disallowDanglingUnderscores
/*eslint-disable no-underscore-dangle*/

module.exports = function (grunt) {
    const commons = require('js-project-commons');

    grunt.loadNpmTasks('grunt-shell');

    commons.grunt.config.initConfig(grunt, {
        buildConfig: {
            projectRoot: __dirname,
            nodeProject: true,
            skipSecurityCheck: true
        },
        apidoc2readme: {
            readme: {
                options: {
                    tags: {
                        'connection-upsert': 'Connection.upsert'
                    },
                    modifySignature(line) {
                        return line.split('Connection.').join('connection.').split('Promise').join('[Promise]');
                    }
                }
            }
        }
    });
};


lib/upsert.js
'use strict';

/**
 * Invoked when an async operation has finished.
 *
 * @callback AsyncCallback
 * @param {Error} [error] - Any possible error
 * @param {Object} [output] - The operation output
 */

/**
 * This class holds all the extended capabilities added the oracledb connection.
 *
 * @author Sagie Gur-Ari
 * @class Connection
 * @public
 */

/**
 * Returns only the required bind params based on the provided SQL.
 *
 * @function
 * @memberof! Connection
 * @private
 * @param {String} sql - The SQL using the bind params
 * @param {Object} bindParams - The bind parameters used to specify the values for the columns, used by all execute operations
 * @param {Object} [options] - The execute options
 * @returns {Object} The specific SQL bind params
 */
function getBindParams(sql, bindParams, options) {
    const operationBindParams = {};

    let keys;
    const lobParams = {};
    if (options && options.lobMetaInfo) {
        keys = Object.keys(options.lobMetaInfo);

        for (let index = 0; index < keys.length; index++) {
            lobParams[options.lobMetaInfo[keys[index]]] = true;
        }
    }

    keys = Object.keys(bindParams);

    for (let index = 0; index < keys.length; index++) {
        const param = keys[index];

        if ((sql.indexOf(':' + param) !== -1) || lobParams[param]) {
            operationBindParams[param] = bindParams[param];
        }
    }

    return operationBindParams;
}

/*eslint-disable valid-jsdoc*/
//jscs:disable jsDoc
/**
 * The UPSERT oracledb extension gets 3 SQL statements.<br>
 * It first queries the database for existing data, based on the output, it either runs INSERT or UPDATE SQL.<br>
 * If it runs the INSERT and it fails on unique constraint, it will also run the UPDATE.<br>
 * The output in the callback is the output of the INSERT/UPDATE operation.
 *
 * @function
 * @memberof! Connection
 * @public
 * @param {Object} sqls - Holds all required SQLs to support the UPSERT capability
 * @param {String} sqls.query - The SELECT SQL statement
 * @param {String} sqls.insert - The INSERT SQL statement
 * @param {String} sqls.update - The UPDATE SQL statement
 * @param {Object} bindParams - The bind parameters used to specify the values for the columns, used by all execute operations (arrays not supported, only named bind params)
 * @param {Object} [options] - Used by all execute operations
 * @param {AsyncCallback} [callback] - Invoked once the UPSERT is done with either an error or the output
 * @returns {Promise} In case of no callback provided in input and promise is supported, this function will return a promise
 * @example
 * ```js
 * connection.upsert({
 *   query: 'SELECT ID FROM MY_DATA WHERE ID = :id',
 *   insert: 'INSERT INTO MY_DATA (ID, NAME) VALUES (:id, :name)',
 *   update: 'UPDATE MY_DATA SET NAME = :name WHERE ID = :id'
 * }, {
 *   id: 110,
 *   name: 'new name'
 * }, {
 *   autoCommit: false
 * }, function onUpsert(error, results) {
 *   if (error) {
 *     //handle error...
 *   } else {
 *     console.log('rows affected: ', results.rowsAffected);
 *
 *     //continue flow...
 *   }
 * });
 * ```
 */
const upsert = function (sqls, bindParams, options, callback) {
    /*eslint-disable no-invalid-this*/
    const self = this;
    /*eslint-enable no-invalid-this*/

    if (!callback) {
        callback = options;
        options = null;
    }

    bindParams = bindParams || {};
    options = options || {};

    if (Array.isArray(bindParams)) {
        callback(new Error('Array type bind params are not supported.'));
    } else {
        let upsertResult;
        let runUpdate;

        self.run([
            function query(asyncCallback) {
                //set/add query options
                options.maxRows = 1;
                options.resultSet = false;

                const operationBindParams = getBindParams(sqls.query, bindParams);

                self.query(sqls.query, operationBindParams, options, function onQueryDone(error, results) {
                    if (error) {
                        asyncCallback(error);
                    } else {
                        if (results && results.length) {
                            runUpdate = true;
                        }

                        asyncCallback();
                    }
                });
            },
            function insert(asyncCallback) {
                if (runUpdate) {
                    asyncCallback();
                } else {
                    const operationBindParams = getBindParams(sqls.insert, bindParams, options);

                    self.insert(sqls.insert, operationBindParams, options, function onInsertDone(error, results) {
                        if (error) {
                            if (error.message && (error.message.indexOf('ORA-00001') !== -1)) { //INSERT failed on unique constraint so try to UPDATE instead
                                runUpdate = true;
                                asyncCallback();
                            } else {
                                asyncCallback(error);
                            }
                        } else {
                            if (results && results.rowsAffected) {
                                upsertResult = results;
                            } else {
                                runUpdate = true;
                            }

                            asyncCallback();
                        }
                    });
                }
            },
            function update(asyncCallback) {
                if (runUpdate) {
                    const operationBindParams = getBindParams(sqls.update, bindParams, options);

                    self.update(sqls.update, operationBindParams, options, function onUpdateDone(error, results) {
                        if (error) {
                            asyncCallback(error);
                        } else {
                            let upsertError;
                            if (results && results.rowsAffected) {
                                upsertResult = results;
                            } else {
                                upsertError = new Error('No rows updated.');
                            }

                            asyncCallback(upsertError);
                        }
                    });
                } else {
                    asyncCallback();
                }
            }
        ], {
            sequence: true
        }, function onUpsertDone(error) {
            callback(error, upsertResult);
        });
    }
};
//jscs:enable jsDoc
/*eslint-enable valid-jsdoc*/

const SimpleOracleDB = require('simple-oracledb');
SimpleOracleDB.addExtension('connection', 'upsert', upsert);
